engine/src/cubos/engine/gl/deferred/renderer.cpp

```diff
@@ -1,465 +1,569 @@
+    vaDesc.elements[0].buffer.stride = sizeof(Vertex);
+    vaDesc.elements[1].name = "normal";
+    vaDesc.elements[1].type = Type::Float;
+    vaDesc.elements[1].size = 3;
+    vaDesc.elements[1].buffer.index = 0;
+    vaDesc.elements[1].buffer.offset = offsetof(Vertex, normal);
+    vaDesc.elements[1].buffer.stride = sizeof(Vertex);
+    vaDesc.elements[2].name = "material";
+    vaDesc.elements[2].type = Type::UShort;
+    vaDesc.elements[2].size = 1;
+    vaDesc.elements[2].buffer.index = 0;
+    vaDesc.elements[2].buffer.offset = offsetof(Vertex, material);
+    vaDesc.elements[2].buffer.stride = sizeof(Vertex);
+    vaDesc.buffers[0] =
+        this->renderDevice.createVertexBuffer(vertices.size() * sizeof(Vertex), vertices.data(), Usage::Static);
+    vaDesc.shaderPipeline = this->geometryPipeline;
+    gpuGrid.va = renderDevice.createVertexArray(vaDesc);
+    gpuGrid.ib =
+        renderDevice.createIndexBuffer(indices.size() * sizeof(size_t), &indices[0], IndexFormat::UInt, Usage::Static);
+    gpuGrid.indexCount = indices.size();
+
+    // Add it to the vector of grids.
+    this->grids.push_back(gpuGrid);
+    return this->grids.size() - 1;
 }
 
-engine::gl::Renderer::ModelID deferred::Renderer::registerModel(const core::gl::Grid& grid)
+void deferred::Renderer::free(GridID grid)
 {
-    registerRequests.emplace_back(grid);
-    return modelCounter++;
+    this->grids[grid].va = nullptr;
+    this->grids[grid].ib = nullptr;
+    this->grids[grid].indexCount = 0;
 }
 
-void deferred::Renderer::drawLight(const SpotLight& light)
+void deferred::Renderer::setPalette(const core::gl::Palette& palette)
 {
-    if (lights.numSpotLights >= CUBOS_DEFERRED_RENDERER_MAX_SPOT_LIGHT_COUNT)
+    // Get the colors from the palette.
+    // Magenta is used for non-existent materials in order to easily identify errors.
+    std::vector<glm::vec4> data(65536, {1.0f, 0.0f, 1.0f, 1.0f});
+    for (size_t i = 0; i < palette.getSize(); ++i)
     {
-        logWarning("deferred::Renderer::drawLight(const SpotLight& light) failed: number of spot lights to be drawn "
-                   "this frame exceeds CUBOS_DEFERRED_RENDERER_MAX_SPOT_LIGHT_COUNT.");
-        return;
+        if (palette.getData()[i].similarity(Material::Empty) < 1.0f)
+        {
+            data[i + 1] = palette.getData()[i].color;
+        }
     }
-    lights.spotLights[lights.numSpotLights++] = light;
-}
 
-void deferred::Renderer::drawLight(const DirectionalLight& light)
-{
-    if (lights.numDirectionalLights >= CUBOS_DEFERRED_RENDERER_MAX_DIRECTIONAL_LIGHT_COUNT)
-    {
-        logWarning("deferred::Renderer::drawLight(const DirectionalLight& light) failed: number of directional "
-                   "lights to be drawn this frame exceeds CUBOS_DEFERRED_RENDERER_MAX_DIRECTIONAL_LIGHT_COUNT.");
-        return;
-    }
-    lights.directionalLights[lights.numDirectionalLights++] = light;
+    // Update the data to the GPU.
+    this->paletteTex->update(0, 0, 256, 256, &data[0]);
```

Comment:
For `std::vector` you can do `data.data()` instead for a cleaner way:
```suggestion
    this->paletteTex->update(0, 0, 256, 256, data.data());
```